mmengine/runner/runner.py

```diff
@@ -2080,14 +2089,28 @@ def save_checkpoint(
         if hasattr(self.train_dataloader.dataset, 'metainfo'):
             meta.update(dataset_meta=self.train_dataloader.dataset.metainfo)
 
-        if is_model_wrapper(self.model):
-            model = self.model.module
+        # models with this attribute are FSDP models, which has special logic
```

Comment:
https://github.com/open-mmlab/mmengine/blob/24c9627f8ada684a00625296c798c513690256cf/mmengine/runner/runner.py#L1949

I'm afraid to leave a comment because I don't know if your work is over. It seems that the optimizer state processing for the fsdp model is necessary.  I marked the suggestion on the line closest to where I can leave a comment.

Refer to the [official PyTorch documentation](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.shard_full_optim_state_dict)

```suggestion
            self.optim_wrapper.load_state_dict(
                FullyShardedDataParallel.shard_full_optim_state_dict(
                    checkpoint['optimizer'], self.model) if isinstance(
                        self.model, FullyShardedDataParallel) else
                checkpoint['optimizer'])
```

Reply:
Assuming that the full optimizer state is stored, sharding the optimizer state dict part was added.